### PR TITLE
Go to Type Declaration via LSP typeDefinition

### PR DESCRIPTION
**Pull request by ralph-bergmann**
Implements #580 with the same pattern as #539/#552 and the inlay-hints PR this is stacked on (#617): the bridge forwards `textDocument/typeDefinition` (advertising `typeDefinitionProvider`), `LspMethod.TYPE_DEFINITION` is added, and `goToTypeDefinitionCustomizer` is enabled behind the experimental-LSP setting. The bundled client's `LspImplicitReferenceProvider`/`LspSymbolTypeProvider` drive *Navigate | Type Declaration*; the PSI path is only consulted when the server returns nothing, so there is nothing to gate.

**Independent of #617** — this branch is based directly on `main` and contains only the three typeDefinition commits. The two PRs overlap textually (capabilities block, `LspMethod`, descriptor imports, CHANGELOG), so whichever merges second gets a trivial rebase. (I first tried to stack this on #617's branch, but GitHub stacked PRs don't support the fork→upstream case yet — github/gh-stack#46 — so the PRs are simply separate.)

**Client capabilities:** the bridge accepts every shape the server can return for `textDocument/typeDefinition` — `List<LocationLink>` (only sent when the client advertises `typeDefinition.linkSupport`), `List<Location>`, or a single `Location` (what DAS returns today without the capability) — so this PR does not depend on the capability JSON. Advertising `linkSupport` for parity with `definition` (it adds the origin range) can follow once #614 has moved the capabilities into `DartAnalysisServerService.buildLspCapabilities`.

**Manual test:** flag on, caret on a variable (not on the type name) → ⌃⇧B / Ctrl+Shift+B (*Navigate | Type Declaration*) jumps to the declaration of the variable's type — verified in the sandbox with a Flutter project, including a type in `.pub-cache` (`GoRouter`). Flag off → unchanged (no Dart implementation existed). No errors in idea.log.

**Functional differences:** none to gate — the plugin never registered a `TypeDeclarationProvider`, so *Go to Type Declaration* did nothing for Dart before. Platform limitation: Ctrl+hover does not underline type-declaration targets for LSP-backed files (`LspImplicitReferenceProvider` only reacts to the explicit action). No SDK version gate: the handler is shared over LSP-over-Legacy since Dart 3.3.0.

**Status:** draft until the remaining sandbox items (file open at startup vs. opened later, `dart:` SDK file, toggle lifecycle) are ticked and screenshots are added. Unit tests (`com.jetbrains.lang.dart.lsp.*`, 11 bridge tests), `verifyPlugin` (no new baseline lines attributable to this change) and the repo code-review skill are done.

---

Review the contribution guidelines below:

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.
- [x] I've included the required information in the description above.
- [x] I've updated `CHANGELOG.md` if appropriate (i.e. user-facing change)

<details>
  <summary>Contribution guidelines:</summary><br>

- See the [Flutter organization contributor guide](https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Dart contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use
  `dart format`.
- Java and Kotlin contributions should strive to follow Java and Kotlin best
  practices ([discussion](https://github.com/flutter/flutter-intellij/issues/8098)).

</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: third_party/src/main/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServer.kt [hunk 1]
```diff
@@ -248,11 +253,25 @@ class DartBridgeLspServer(private val project: Project) : DartLanguageServer, Te
...
     }
 
+    // Note: We advertise linkSupport: true for typeDefinition in server.setClientCapabilities
+    // (see DartAnalysisServerService.buildLspCapabilities) so DAS returns List<LocationLink>.
+    override fun typeDefinition(params: TypeDefinitionParams): CompletableFuture<Either<List<Location>, List<LocationLink>>> {
+        val type = object : TypeToken<List<LocationLink>>() {}.type
+        return forwardRequest<List<LocationLink>>("textDocument/typeDefinition", params, type).thenApply { links ->
+            Either.forRight(links ?: emptyList())
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ralph-bergmann) [reply third_party/src/main/java/com/jetbrains/lang/dart/lsp/DartBridgeLspServer.kt]: Agreed, and done in 7519c92d — with one refinement over the suggestion: without `typeDefinition.linkSupport` the server returns a **single `Location` object** (`Definition.t2(location)` in `handler_type_definition.dart`), not a list, so treating a non-array result as "empty" would have dropped the only hit. The override now reads the raw result and maps all shapes onto the lsp4j `Either`: `[LocationLink…]` → right, `[Location…]` → left, bare `Location` → left with one element, `null`/`[]` → empty. Three new tests cover the Location shapes (`testTypeDefinitionRequestWithSingleLocationResult`, `…WithLocationListResult`, `…WithEmptyResult`). As a consequence this PR no longer depends on #614 for functionality — advertising `linkSupport` later would only add the origin range, as for `definition`. Verified in the sandbox: with the experimental flag on, ⌃⇧B on a variable jumps to its type's class (also into `.pub-cache`); with the flag off nothing happens (previous behaviour); no errors in idea.log.

**Comment by ralph-bergmann** [comment]
Closing this one: it duplicates #615, which was opened the evening before I started — sorry for the noise, I should have refreshed the list of open PRs before picking this up, and I'll do that from now on.

#615 is the better fit anyway: it enables Go to Type Declaration globally (there is no legacy path to fall back to), which makes sense to me. The only things this PR had on top are the bridge tolerating `Location`/bare-`Location` results (not needed once `typeDefinition.linkSupport` is advertised, as #615 does) and a few extra bridge tests (null result, sent params); if any of that is useful I'm happy to add it to #615 as a follow-up. Branch stays around for a while in case you want to pick something from it.
